### DISCUSSION TITLE
Is there a way to search for duplicate files inside archives with same directory structure

### DISCUSSION THREAD

**Discussion by sailukar** [dnGrep/dnGrep#1233] [Q&A] [answered]
Iam trying to find java class with same package structure like below, 

I want to find and list the x.class in the below example

first.jar       com/a/b/c/x.class
second.jar  com/a/b/c/x.class

**Comment by doug24** [comment]
Assuming the class uses the java convention of naming the package to match the directory structure, you can search for: `com/a/b/c/x` without the .class extension.  You will probably get the source class, and other classes that import it - I couldn't find any way around that.

Some things to note
`.class` files are binary
Search in archives - checked
Patterns to match - *.class
Include binary files - checked
Encoding - UTF-8
Set the encoding because dnGrep cannot detect the encoding of binary files. UTF-8 appears to work for strings in .class files.


![image](https://github.com/user-attachments/assets/a97bf03d-80a6-4a9c-a773-eef9ba13e754)

**Reply by sailukar** [reply]
Thanks for the quick response.
Yes this will work if i already know the duplicate folder name or file name, but i dont have any details to search for.
Just wanted to ask as i too find it not possible currently, but wanted to know as i am doing it frequently.

**Comment by doug24** [comment, marked as answer]
OK, I don't know of any easy way to do this, but here is something you can try.  Release 4.2.46 added a new feature to search for a list of strings from an input file:

> #1142 **Use file as search input**: Added a new feature where you can use a file to search for many patterns and generate one result set. 
> This works with regular expressions or text search. In the 'Search for' field enter the full path to a local input file with the `file://` URL prefix like this:
> ```
> file://C:\testFiles\issue1142\searchfor.txt
> ```
> The input file is a plain text file. Put each search pattern or search text on a separate line in the file. Empty lines will be skipped, but no comments or other text are allowed. dnGrep will search each file with all of the search patterns and combine the matches found in the results.
> 

You would start with the files in first.jar and generate a list of strings for each class in the jar, and output that to a file.  Then search second.jar with that file as input.

You can generate a list of files in a jar using dnGrep by leaving the "Search for" pattern empty.  Create the file by doing More -> Save results -> Text results. That will give you the list of class files with the full path of first.jar on the front of each line.  Open this file in your favorite editor and search/replace the jar file path with empty string.  Search/replace .class with empty string.  Search/replace all the backslashes with forward slashes. That should leave the file names in the right format.

If your jars have a lot of classes, this is going to be a very long search.
